### ISSUE TITLE
Test failure: //backends/gpu/cpp_tests:event_manager_test

### ISSUE THREAD

**Issue by tlemo** [tensorflow/runtime#37]
The GCC + NVCC build shows a failure under `//backends/gpu/...`:

```
bazel test --config=cuda --config=gcc //backends/gpu/...
...
INFO: Elapsed time: 1348.192s, Critical Path: 952.08s
INFO: 4412 processes: 55 internal, 4357 linux-sandbox.
INFO: Build completed, 1 test FAILED, 4412 total actions
//backends/gpu/cpp_tests:memory/block_allocator_test                     PASSED in 0.4s
//backends/gpu/cpp_tests:memory/gpu_buffer_test                          PASSED in 0.2s
//backends/gpu/cpp_tests:memory/sub_allocator_test                       PASSED in 0.1s
//backends/gpu/cpp_tests:module_table_test                               PASSED in 0.9s
//backends/gpu/cpp_tests:stream/wrapper_test                             PASSED in 3.5s
//backends/gpu/mlir_tests/core_runtime:current_context_lifetime.mlir.test PASSED in 6.6s
//backends/gpu/mlir_tests/core_runtime:get_device.mlir.test              PASSED in 1.5s
//backends/gpu/mlir_tests/core_runtime:logging_with_results.mlir.test    PASSED in 1.5s
//backends/gpu/mlir_tests/core_runtime:op_handler.mlir.test              PASSED in 1.5s
//backends/gpu/mlir_tests/core_runtime:tensor_conversion.mlir.test       PASSED in 1.5s
//backends/gpu/mlir_tests/core_runtime:test_ops.mlir.test                PASSED in 6.5s
//backends/gpu/mlir_tests/core_runtime:tf_addv2.mlir.test                PASSED in 6.6s
//backends/gpu/mlir_tests/core_runtime:tf_biasadd.mlir.test              PASSED in 1.5s
//backends/gpu/mlir_tests/core_runtime:tf_cast.mlir.test                 PASSED in 6.5s
//backends/gpu/mlir_tests/core_runtime:tf_const.mlir.test                PASSED in 1.5s
//backends/gpu/mlir_tests/core_runtime:tf_conv2d.mlir.test               PASSED in 7.3s
//backends/gpu/mlir_tests/core_runtime:tf_fusedbatchnorm.mlir.test       PASSED in 6.6s
//backends/gpu/mlir_tests/core_runtime:tf_matmul.mlir.test               PASSED in 6.5s
//backends/gpu/mlir_tests/core_runtime:tf_maxpool.mlir.test              PASSED in 1.5s
//backends/gpu/mlir_tests/core_runtime:tf_mean.mlir.test                 PASSED in 0.5s
//backends/gpu/mlir_tests/core_runtime:tf_pad.mlir.test                  PASSED in 6.5s
//backends/gpu/mlir_tests/core_runtime:tf_reduction_ops.mlir.test        PASSED in 6.6s
//backends/gpu/mlir_tests/core_runtime:tf_relu.mlir.test                 PASSED in 6.5s
//backends/gpu/mlir_tests/core_runtime:tf_softmax.mlir.test              PASSED in 1.5s
//backends/gpu/mlir_tests/core_runtime:tf_tanh.mlir.test                 PASSED in 6.7s
//backends/gpu/mlir_tests/core_runtime:tf_transpose.mlir.test            PASSED in 6.6s
//backends/gpu/mlir_tests/cuda:cublas.mlir.test                          PASSED in 1.0s
//backends/gpu/mlir_tests/cuda:dnn.mlir.test                             PASSED in 1.6s
//backends/gpu/mlir_tests/cuda:event.mlir.test                           PASSED in 0.4s
//backends/gpu/mlir_tests/cuda:launch.mlir.test                          PASSED in 0.5s
//backends/gpu/mlir_tests/cuda:memcpy.mlir.test                          PASSED in 0.4s
//backends/gpu/mlir_tests/cuda:module.mlir.test                          PASSED in 0.5s
//backends/gpu/mlir_tests/cuda:stream.mlir.test                          PASSED in 0.5s
//backends/gpu/cpp_tests:event_manager_test                              FAILED in 227.8s
  /home/lemo/.cache/bazel/_bazel_lemo/b9a83ea24871c2ac3b49c420e1a458ae/execroot/tf_runtime/bazel-out/k8-opt/testlogs/backends/gpu/cpp_tests/event_manager_test/test.log
Executed 34 out of 34 tests: 33 tests pass and 1 fails locally.
```

Configuration details: GCC 8.4, CUDA 11.1, Ubuntu 18.04 (x64). The GPU is a Titan RTX.

**Comment by tlemo** [comment]
Running the failed test indicates a hard crash:

```
Running bazel-bin/backends/gpu/cpp_tests/event_manager_test
Run on (32 X 3400 MHz CPU s)
CPU Caches:
  L1 Data 32K (x16)
  L1 Instruction 64K (x16)
  L2 Unified 512K (x16)
  L3 Unified 8192K (x4)
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
I 2021-03-18 19:06:28.259879 139907117381376 event_manager.cc:82] EventManager worker thread cancelled.
I 2021-03-18 19:06:28.259999 139907283670976 event_manager.cc:154] EventManager destroyed with 0 pending events.
I 2021-03-18 19:06:28.373243 139907117381376 event_manager.cc:82] EventManager worker thread cancelled.
I 2021-03-18 19:06:28.373338 139907283670976 event_manager.cc:154] EventManager destroyed with 0 pending events.
I 2021-03-18 19:06:28.468926 139907117381376 event_manager.cc:82] EventManager worker thread cancelled.
I 2021-03-18 19:06:28.469015 139907283670976 event_manager.cc:154] EventManager destroyed with 0 pending events.
I 2021-03-18 19:06:28.555418 139907117381376 event_manager.cc:82] EventManager worker thread cancelled.
I 2021-03-18 19:06:28.555507 139907283670976 event_manager.cc:154] EventManager destroyed with 0 pending events.
I 2021-03-18 19:06:28.654265 139907117381376 event_manager.cc:82] EventManager worker thread cancelled.
I 2021-03-18 19:06:28.654355 139907283670976 event_manager.cc:154] EventManager destroyed with 0 pending events.
I 2021-03-18 19:06:28.762471 139907117381376 event_manager.cc:82] EventManager worker thread cancelled.
I 2021-03-18 19:06:28.762561 139907283670976 event_manager.cc:154] EventManager destroyed with 0 pending events.
I 2021-03-18 19:06:29.017772 139907117381376 event_manager.cc:82] EventManager worker thread cancelled.
I 2021-03-18 19:06:29.017860 139907283670976 event_manager.cc:154] EventManager destroyed with 0 pending events.
I 2021-03-18 19:06:29.818921 139907117381376 event_manager.cc:82] EventManager worker thread cancelled.
I 2021-03-18 19:06:29.819012 139907283670976 event_manager.cc:154] EventManager destroyed with 0 pending events.
-----------------------------------------------------------------------------
Benchmark                                      Time           CPU Iterations
-----------------------------------------------------------------------------
BM_EventEnqueue                              161 ns        161 ns    4399306
I 2021-03-18 19:06:29.911051 139907100595968 event_manager.cc:82] EventManager worker thread cancelled.
I 2021-03-18 19:06:29.911129 139907283670976 event_manager.cc:154] EventManager destroyed with 0 pending events.
I 2021-03-18 19:06:30.011420 139907100595968 event_manager.cc:82] EventManager worker thread cancelled.
I 2021-03-18 19:06:30.011515 139907283670976 event_manager.cc:154] EventManager destroyed with 0 pending events.
I 2021-03-18 19:06:30.102653 139907100595968 event_manager.cc:82] EventManager worker thread cancelled.
I 2021-03-18 19:06:30.102724 139907283670976 event_manager.cc:154] EventManager destroyed with 0 pending events.
I 2021-03-18 19:06:30.219432 139907100595968 event_manager.cc:82] EventManager worker thread cancelled.
I 2021-03-18 19:06:30.219521 139907283670976 event_manager.cc:154] EventManager destroyed with 0 pending events.
I 2021-03-18 19:06:30.487184 139907100595968 event_manager.cc:82] EventManager worker thread cancelled.
I 2021-03-18 19:06:30.487267 139907283670976 event_manager.cc:154] EventManager destroyed with 0 pending events.
I 2021-03-18 19:06:31.845045 139907100595968 event_manager.cc:82] EventManager worker thread cancelled.
I 2021-03-18 19:06:31.845123 139907283670976 event_manager.cc:154] EventManager destroyed with 0 pending events.
...
BM_MoreStreamsThanThreads/1/10            860457 ns     111927 ns       5371
I 2021-03-18 19:10:23.208142 139906512828160 event_manager.cc:82] EventManager worker thread cancelled.
I 2021-03-18 19:10:23.208226 139907283670976 event_manager.cc:154] EventManager destroyed with 0 pending events.
I 2021-03-18 19:10:23.321154 139906345072384 event_manager.cc:82] EventManager worker thread cancelled.
I 2021-03-18 19:10:23.321245 139907283670976 event_manager.cc:154] EventManager destroyed with 0 pending events.
I 2021-03-18 19:10:23.512955 139906496059136 event_manager.cc:82] EventManager worker thread cancelled.
I 2021-03-18 19:10:23.513032 139907283670976 event_manager.cc:154] EventManager destroyed with 0 pending events.
I 2021-03-18 19:10:24.461173 139906496059136 event_manager.cc:82] EventManager worker thread cancelled.
I 2021-03-18 19:10:24.461241 139907283670976 event_manager.cc:154] EventManager destroyed with 0 pending events.
I 2021-03-18 19:10:28.868175 139906496059136 event_manager.cc:82] EventManager worker thread cancelled.
I 2021-03-18 19:10:28.868265 139907283670976 event_manager.cc:154] EventManager destroyed with 0 pending events.
BM_MoreStreamsThanThreads/8/10            864058 ns      90383 ns       4971
Segmentation fault (core dumped)
```

**Comment by tlemo** [comment]
From a quick glance, it seems that the entries in `tfrt::HostContext::all_host_contexts_` are never freed, and `next_host_context_index` eventually overflows

**Comment by jing-dong** [comment]
Thanks for reporting this issue. It should have been fixed in commit d1d7f44f.

**Comment by tlemo** [comment]
I can confirm that the test case now passes.

```
//backends/gpu/cpp_tests:event_manager_test                              PASSED in 239.1s
```
